go/adapter/datadog/adapter.go

```diff
@@ -43,76 +39,54 @@ func NewDatadogAdapter(config *DatadogConfig) (DatadogAdapter, error) {
 
 	return DatadogAdapter{
 		AdapterBase: observe.NewAdapterBase(),
-		TraceId:     uint64(observe.NewTraceId()),
 		Config:      config,
 	}, nil
 }
 
-func (d *DatadogAdapter) Event(e observe.Event) {
-	switch event := e.(type) {
-	case observe.CallEvent:
-		spans := d.makeCallSpans(event, nil)
-		if len(spans) > 0 {
-			d.Spans = append(d.Spans, spans...)
-		}
-
-	case observe.MemoryGrowEvent:
-		if len(d.Spans) > 0 {
-			d.Spans[len(d.Spans)-1].AddAllocation(event.MemoryGrowAmount())
-		}
-	case observe.CustomEvent:
-		if value, ok := event.Metadata["trace_id"]; ok {
-			traceId, err := strconv.ParseUint(value.(string), 10, 64)
-			if err != nil {
-				log.Println("failed to parse traceId from event metadata:", err)
-				return
-			}
-
-			d.TraceId = traceId
-		}
-	}
+func (d *DatadogAdapter) Start() {
+	d.AdapterBase.Start(d)
 }
 
-func (d *DatadogAdapter) Wait(collector *observe.Collector, timeout time.Duration) {
-	d.AdapterBase.Wait(collector, timeout, nil)
+func (d *DatadogAdapter) Stop() {
+	d.AdapterBase.Stop()
 }
 
-func (d *DatadogAdapter) Start(collector *observe.Collector, wasm []byte) error {
-	if err := d.AdapterBase.Start(collector, wasm); err != nil {
-		return err
+func (d *DatadogAdapter) HandleTraceEvent(te observe.TraceEvent) {
+	if te.TelemetryId == nil {
+		log.Println("Datadog adapter needs a trace id")
```

Comment:
should we generate a TelemetryId in the event one isn't supplied? Or is that generated value supposed to be taken care of already prior to this point?

Reply:
default is applied upstream. if it's not here now then that's a problem so i just log and return

Reply:
we could maybe just make it non-nillable too.

Reply:
> default is applied upstream

ok, then this is probably fine as-is.. if we get into this scenario a log will be helpful as some info from a customer trying to debug. 

> we could maybe just make it non-nillable too.

yea, we could do that but it would still be possible for the value to be `0` by default. In Go a new struct is initialized with the field default values no matter what. So we'd end up checking for `0` too?

Reply:
I think I'll make the property a TelemetryId and it will just always be present. if the adapter doesn't need it it can ignore it. Otherwise the type system will make sure there is one there.